File: objective-c/ArcGISSample/Time/TimeViewController.m
//
//  TimeViewController.m
//  ArcGISSample
//
//  Created by esrij on 2015/07/27.
//  Copyright (c) 2015年 esrij. All rights reserved.
//

#import "TimeViewController.h"


@implementation TimeViewController



- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.agsMapView = [[AGSMapView alloc] initWithFrame:self.view.bounds];
	[self.view addSubview:self.agsMapView];
    
    //タイルマップサービスレイヤーの追加
    NSURL *url = [NSURL URLWithString:@"http://server.arcgisonline.com/ArcGIS/rest/services/World_Topo_Map/MapServer"];
    AGSTiledMapServiceLayer *tiledMapServiceLayer = [[AGSTiledMapServiceLayer alloc] initWithURL:url];
    [self.agsMapView addMapLayer:tiledMapServiceLayer withName:@"Tiled Layer"];

    //時間対応レイヤー表示用のフィーチャレイヤーの追加
    NSURL *timeUrl = [NSURL URLWithString:@"http://sampleserver3.arcgisonline.com/ArcGIS/rest/services/Hurricanes/NOAA_Tracks_1851_2007/MapServer/0"];
    AGSFeatureLayer *agsFeatureLayer = [[AGSFeatureLayer alloc] initWithURL:timeUrl mode:AGSFeatureLayerModeOnDemand];
	[self.agsMapView addMapLayer:agsFeatureLayer withName:@"Time Layer"];
    
    //マップ上でフィーチャを表示する時間範囲を指定
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy-MM-dd"];
    
    self.startDate = [dateFormatter dateFromString:@"1900-1-1"];
    self.endDate = [dateFormatter dateFromString:@"1901-1-1"];
    
    AGSTimeExtent *agsTimeExtent = [[AGSTimeExtent alloc] initWithStart:self.startDate end:self.endDate];
    self.agsMapView.timeExtent = agsTimeExtent;
    
    
    UIBarButtonItem *timeItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemPlay target:self action:@selector(buttonEvent:)];
    self.navigationItem.rightBarButtonItem = timeItem;
    
}


- (void)buttonEvent:(UIButton *)button {
    
    //マップ上でフィーチャを表示する時間範囲を1年単位で変更
    NSDateComponents *comps = [[NSDateComponents alloc] init];
    comps.year = 1;
    NSCalendar *calendar = [NSCalendar currentCalendar];
    
    NSDate *startDateNew = [calendar dateByAddingComponents:comps toDate:self.startDate options:0];
    NSDate *endDateNew = [calendar dateByAddingComponents:comps toDate:self.endDate options:0];
    
    self.startDate = startDateNew;
    self.endDate = endDateNew;
    
    NSLog(@"Start:%@, End:%@", [self.startDate description], [self.endDate description]);
    
    AGSTimeExtent *agsTimeExtent = [[AGSTimeExtent alloc] initWithStart:self.startDate end:self.endDate];
    self.agsMapView.timeExtent = agsTimeExtent;
    
}


- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: swift/ArcGISSample/TimeViewController.swift
//
//  TimeViewController.swift
//  ArcGISSample
//
//  Created by esrij on 2015/07/27.
//  Copyright (c) 2015年 esrij. All rights reserved.
//

import UIKit
import ArcGIS


class TimeViewController: UIViewController, AGSWebMapDelegate {
    
    
    var agsMapView: AGSMapView!
    
    var startDate: Date!
    var endDate: Date!

    
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        agsMapView = AGSMapView(frame: view.bounds)
        view.addSubview(agsMapView)
        
        //タイルマップサービスレイヤーの追加
        let url = URL(string: "https://server.arcgisonline.com/ArcGIS/rest/services/World_Topo_Map/MapServer")
        let tiledLyr = AGSTiledMapServiceLayer(url:url)
        agsMapView.addMapLayer(tiledLyr, withName:"Tiled Layer")
        
        //時間対応レイヤー表示用のフィーチャレイヤーの追加
        let timeUrl = URL(string: "https://sampleserver3.arcgisonline.com/ArcGIS/rest/services/Hurricanes/NOAA_Tracks_1851_2007/MapServer/0")
        let agsFeatureLayer = AGSFeatureLayer(url: timeUrl, mode: .onDemand)
        agsMapView.addMapLayer(agsFeatureLayer, withName:"Time Layer")

        //マップ上でフィーチャを表示する時間範囲を指定
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        
        startDate = dateFormatter.date(from: "1900-1-1")
        endDate = dateFormatter.date(from: "1901-1-1")
        
        let agsTimeExtent = AGSTimeExtent(start: startDate, end: endDate)
        agsMapView.timeExtent = agsTimeExtent
        
        
        let timeItem = UIBarButtonItem(barButtonSystemItem: .play, target: self, action: #selector(TimeViewController.buttonEvent))
        navigationItem.rightBarButtonItem = timeItem
        
    }
    
    
    func buttonEvent(_ sender: UIBarButtonItem) {
        
        //マップ上でフィーチャを表示する時間範囲を1年単位で変更
        var comps = DateComponents()
        comps.year = 1
        let calendar = Calendar.current
        
        let startDateNew = calendar.date(byAdding: comps, to: startDate)
        let endDateNew = calendar.date(byAdding: comps, to: endDate)

        
        startDate = startDateNew
        endDate = endDateNew
        
        print("Start:\(startDate.description)", "End:\(endDate.description)")

        let agsTimeExtent = AGSTimeExtent(start: startDate, end: endDate)
        agsMapView.timeExtent = agsTimeExtent
        
    }
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
}
